module_utils/azure_rm_common.py

```diff
@@ -345,6 +376,12 @@ def __init__(self, derived_arg_spec, bypass_checks=False, no_log=False,
             res = self.exec_module(**self.module.params)
             self.module.exit_json(**res)
 
+    def acquire_token_with_username_password(self, authority, resource, username, password, client_id):
+        context = AuthenticationContext(authority)
```

Comment:
Will this raise any exception? Like the credential is invalid or just without granted any access. If so, should catch the exception and give a friendly message

Reply:
ServicePrincipleCredential, and UserPassCredential don't handle exception, just throw them out. Seems there's no auth exception handling in original code. I follow that practice, and tested below exception scenario, exception looks like:
---
An exception occurred during task execution. To see the full traceback, use -vvv. The error was: Timestamp: 2018-03-16 03:25:48Z
fatal: [localhost]: FAILED! => {"changed": false, "module_stderr": "Traceback (most recent call last):\n  File \"/home/yungez/src/ansible/venv/lib/python3.5/site-packages/msrestazure/azure_active_directory.py\", line 402, in set_token\n    **optional)\n  File \"/home/yungez/src/ansible/venv/lib/python3.5/site-packages/requests_oauthlib/oauth2_session.py\", line 244, in fetch_token\n    self._client.parse_request_body_response(r.text, scope=self.scope)\n  File \"/home/yungez/src/ansible/venv/lib/python3.5/site-packages/oauthlib/oauth2/rfc6749/clients/base.py\", line 408, in parse_request_body_response\n    self.token = parse_token_response(body, scope=scope)\n  File \"/home/yungez/src/ansible/venv/lib/python3.5/site-packages/oauthlib/oauth2/rfc6749/parameters.py\", line 379, in parse_token_response\n    validate_token_parameters(params)\n  File \"/home/yungez/src/ansible/venv/lib/python3.5/site-packages/oauthlib/oauth2/rfc6749/parameters.py\", line 386, in validate_token_parameters\n    raise_from_error(params.get('error'), params)\n  File \"/home/yungez/src/ansible/venv/lib/python3.5/site-packages/oauthlib/oauth2/rfc6749/errors.py\", line 415, in raise_from_error\n    raise cls(**kwargs)\noauthlib.oauth2.rfc6749.errors.InteractionRequired: (interaction_required) AADSTS50079: The user is required to use multi-factor authentication.\r\nTrace ID: 4d5c0884-687a-40bf-a40c-a26db0d50100\r\nCorrelation ID: 0e96cf89-472b-4cc0-a2b6-170717bfbd08\r\nTimestamp: 2018-03-16 03:25:48Z\n\nDuring handling of the above exception, another exception occurred:\n\nTraceback (most recent call last):\n  File \"/tmp/ansible_9ai705l8/ansible_module_azure_rm_containerregistry.py\", line 429, in <module>\n    main()\n  File \"/tmp/ansible_9ai705l8/ansible_module_azure_rm_containerregistry.py\", line 426, in main\n    AzureRMContainerRegistry()\n  File \"/tmp/ansible_9ai705l8/ansible_module_azure_rm_containerregistry.py\", line 256, in __init__\n    supports_tags=True)\n  File \"/tmp/ansible_9ai705l8/ansible_modlib.zip/ansible/module_utils/azure_rm_common.py\", line 352, in __init__\n  File \"/home/yungez/src/ansible/venv/lib/python3.5/site-packages/msrestazure/azure_active_directory.py\", line 367, in __init__\n    self.set_token()\n  File \"/home/yungez/src/ansible/venv/lib/python3.5/site-packages/msrestazure/azure_active_directory.py\", line 404, in set_token\n    raise_with_traceback(AuthenticationError, \"\", err)\n  File \"/home/yungez/src/ansible/venv/lib/python3.5/site-packages/msrest/exceptions.py\", line 45, in raise_with_traceback\n    raise error.with_traceback(exc_traceback)\n  File \"/home/yungez/src/ansible/venv/lib/python3.5/site-packages/msrestazure/azure_active_directory.py\", line 402, in set_token\n    **optional)\n  File \"/home/yungez/src/ansible/venv/lib/python3.5/site-packages/requests_oauthlib/oauth2_session.py\", line 244, in fetch_token\n    self._client.parse_request_body_response(r.text, scope=self.scope)\n  File \"/home/yungez/src/ansible/venv/lib/python3.5/site-packages/oauthlib/oauth2/rfc6749/clients/base.py\", line 408, in parse_request_body_response\n    self.token = parse_token_response(body, scope=scope)\n  File \"/home/yungez/src/ansible/venv/lib/python3.5/site-packages/oauthlib/oauth2/rfc6749/parameters.py\", line 379, in parse_token_response\n    validate_token_parameters(params)\n  File \"/home/yungez/src/ansible/venv/lib/python3.5/site-packages/oauthlib/oauth2/rfc6749/parameters.py\", line 386, in validate_token_parameters\n    raise_from_error(params.get('error'), params)\n  File \"/home/yungez/src/ansible/venv/lib/python3.5/site-packages/oauthlib/oauth2/rfc6749/errors.py\", line 415, in raise_from_error\n    raise cls(**kwargs)\nmsrest.exceptions.AuthenticationError: , InteractionRequired: (interaction_required) AADSTS50079: The user is required to use multi-factor authentication.\r\nTrace ID: 4d5c0884-687a-40bf-a40c-a26db0d50100\r\nCorrelation ID: 0e96cf89-472b-4cc0-a2b6-170717bfbd08\r\nTimestamp: 2018-03-16 03:25:48Z\n", "module_stdout": "", "msg": "MODULE FAILURE", "rc": 1}
---

Reply:
well, exception handling doesn't work particularly well in Ansible. in general everything should be handled gracefully.
Even in case of import it says "# This is handled in azure_rm_common" but it is not handled gracefully at all.
So i think at some point we should improve authorisation errors handling, as this is going to be a major painpoint.
I think we shoudl creaate an issue for that, and also for import handling.